environ_internal_test.go

```diff
@@ -0,0 +1,52 @@
+package environ
+
+import (
+	"testing"
+)
+
+// TestLocks checks for locking functions missing the second ()
+// in their defer call.
+func TestLocks(t *testing.T) {
+	a := New([]string{"A=A"})
+	fl := fakeLocker{}
+	a.l = &fl
+
+	_ = a.Len()
+	_ = a.Keys()
+	_ = a.AsSlice()
+	_ = a.AsMap()
+	_ = a.Keep("A")
+	_ = a.Drop("A")
+	_, _ = a.MarshalJSON()
+	_ = a.Get("A")
+	a.Set("B", "B")
+	a.Unset("B")
+
+	if fl.locks != 0 {
+		t.Errorf("locks was non-zero %d, search for 'writeLocker\\(\\)$' and add additional parens", fl.locks)
+	}
+
+	if fl.rlocks != 0 {
+		t.Errorf("rlocks was non-zero %d, search for 'readLocker\\(\\)$' and add additional parens", fl.rlocks)
+	}
+}
+
+type fakeLocker struct {
```

Comment:
just a comment, I haven't seen this fake pattern used to test a mutex before - very neat

Reply:
This is why private interfaces are sometimes useful. Implementation of the interface is be able to test the locking is complete by replacing a real lock with this and checking lock count after running all locking functions.